README.md
[![PyPI version](https://badge.fury.io/py/authy.svg)](https://pypi.org/project/authy/) [![Build Status](https://travis-ci.org/twilio/authy-python.svg?branch=master)](https://travis-ci.org/authy/authy-python)

# Python Client for Twilio Authy Two-Factor Authentication (2FA) API

Documentation for Python usage of the Authy API lives in the [official Twilio documentation](https://www.twilio.com/docs/authy/api/).

The Authy API supports multiple channels of 2FA:
* One-time passwords via SMS and voice.
* Soft token ([TOTP](https://www.twilio.com/docs/glossary/totp) via the Authy App)
* Push authentication via the Authy App

If you only need SMS and Voice support for one-time passwords, we recommend using the [Twilio Verify API](https://www.twilio.com/docs/verify/api) instead. 

[More on how to choose between Authy and Verify here.](https://www.twilio.com/docs/verify/authy-vs-verify)

### Authy Quickstart

For a full tutorial, check out either of the Python Authy Quickstarts in our docs:
* [Python/Django Authy Quickstart](https://www.twilio.com/docs/authy/quickstart/two-factor-authentication-python-django)
* [Python/Flask Authy Quickstart](https://www.twilio.com/docs/authy/quickstart/two-factor-authentication-python-flask)

## Authy Python Installation

Install with [pip](https://www.twilio.com/docs/usage/quickstart/devenvironment-python#how-to-install-pip):

    $ pip install authy

**OR**

Download the [source code](https://github.com/twilio/authy-python/archive/master.zip) and run the following command from your terminal:

    $ python setup.py install

_Note that you may need admin permissions to run the above commands._

## Usage

To use the Authy client, import AuthyApiClient and initialize it with your production API Key found in the [Twilio Console](https://www.twilio.com/console/authy/applications/):

```python
from authy.api import AuthyApiClient
authy_api = AuthyApiClient('your_api_key')
```

![authy api key in console](https://s3.amazonaws.com/com.twilio.prod.twilio-docs/images/account-security-api-key.width-800.png)

## 2FA Workflow

1. [Create a user](https://www.twilio.com/docs/authy/api/users#enabling-new-user)
2. [Send a one-time password](https://www.twilio.com/docs/authy/api/one-time-passwords)
3. [Verify a one-time password](https://www.twilio.com/docs/authy/api/one-time-passwords#verify-a-one-time-password)

**OR**

1. [Create a user](https://www.twilio.com/docs/authy/api/users#enabling-new-user)
2. [Send a push authentication](https://www.twilio.com/docs/authy/api/push-authentications)
3. [Check a push authentication status](https://www.twilio.com/docs/authy/api/push-authentications#check-approval-request-status)


## <a name="phone-verification"></a>Phone Verification

[Phone verification now lives in the Twilio API](https://www.twilio.com/docs/verify/api) and has [Python support through the official Twilio helper libraries](https://www.twilio.com/docs/libraries/python). 

[Legacy (V1) documentation here.](verify-legacy-v1.md) **Verify V1 is not recommended for new development. Please consider using [Verify V2](https://www.twilio.com/docs/verify/api)**.

## Contributing

Install development dependencies with pip:

    sudo pip install -r requirements.txt

To run tests:

    make test
or

    make testfile tests/<test_case_file>

## Copyright

Copyright (c) 2011-2020 Authy Inc. See [LICENSE](https://github.com/twilio/authy-python/blob/master/LICENSE) for further details.


CHANGES
python-authy Changelog
=======================

Here you can see the full list of changes between each python-authy release.

Version 2.2.6
--------------------
[Full Changelog](https://github.com/authy/authy-python/compare/v2.2.5...v2.2.6)

 - [#49](https://github.com/authy/authy-python/pull/49): Remove simplejson as a dependency for python 2.6+

Version 2.2.5
--------------------
[Full Changelog](https://github.com/authy/authy-python/compare/v2.2.4...v2.2.5)

 - [#45](https://github.com/authy/authy-python/pull/45): add registration status endpoint
 - [#47](https://github.com/authy/authy-python/pull/47): Remove reliance on sandbox API for testing, replace with mocks
 - [#44](https://github.com/authy/authy-python/pull/44): Update long_description for PyPI description integration

Version 2.2.4
--------------------
[Full Changelog](https://github.com/authy/authy-python/compare/v2.2.3...v2.2.4)

 - [#43](https://github.com/authy/authy-python/pull/43): Add support for generating generic QR code

Version 2.2.3
--------------------
[Full Changelog](https://github.com/authy/authy-python/compare/v2.2.2...v2.2.3)

 - [4369b12](https://github.com/twilio/authy-python/commit/e8b5856fb3657e2d67373e135642aa18db1e968a): Upgrade requests dependency version

Version 2.2.2
--------------------
[Full Changelog](https://github.com/authy/authy-python/compare/v2.2.1...v2.2.2)

 - [#41](https://github.com/authy/authy-python/pull/41): Coerced token to string to allow for length comparison
 - [#40](https://github.com/authy/authy-python/pull/40): Fix duplicate in CHANGES

Version 2.2.1
--------------------
[Full Changelog](https://github.com/authy/authy-python/compare/v2.2.0...v2.2.1)

 - [#39](https://github.com/authy/authy-python/pull/39): Adds a release script for automating deployment to PyPI
 - [#36](https://github.com/authy/authy-python/pull/36): Add code_length support for phones.verification_start
 - [#24](https://github.com/authy/authy-python/pull/24): Add Python 3 to language classifiers in setup.py

Version 2.2.0
-------------

- Fixes OneTouch signature validation
- Now supports Travis CI so we can release more often 🎉

Version 2.1.5
-------------

- Added onetouch support.
- Request token via call

Version 2.1.4
-------------

- Added make target to run independent test files.

Version 2.1.3
-------------

- Supporting "locale" parameter for Phone Verification. Supported languages are 'en', 'es', 'pt', 'it', 'fr', 'de'.
- Move phone intelligence doc to its own section.
- Improvements to README.
- Updating copyright year.

Version 2.1.2
-------------

- Add Authy User-Agent to requests headers.
- Add method to get current library version.


Version 2.1.1
-------------

- Add support for delete user endpoint.
- Add method to verify if an SMS request was ignored.

Version 2.1.0
-------------

- Add support for fetching user status.
- Add support for start and check phone verification.
- Add support for get phone info.

Version 2.0.0
-------------

- Make use of requests dependency instead of httplib2.

Version 1.1.0
-------------

- Add support for fetching application details.
- Add support for fetching application statistics.

Version 1.0.0
-------------

- Force token verification by default

Version 0.0.4
-------------

- Improve error parsing

Version 0.0.3
-------------

- Bugfixes
- Allow to pass additional params to request_sms and verify token

VERSION 0.0.2
-------------

- Bugfixes
- Support for requesting sms tokens

Version 0.0.1
-------------

- Initial version


test.py
from authy.api import AuthyApiClient

# Your API key from twilio.com/console/authy/applications
# DANGER! This is insecure. See http://twil.io/secure
authy_api = AuthyApiClient('d3rBV0Vd7NIINgoO4Ooc4IyAk8ZjF71L')

authy_id = '57528889'


qr = authy_api.users.generate_qr(authy_id, size='300', label="foo")

print(dir(qr))
print(qr.content['qr_code'])
print(qr.errors())

# sms = authy_api.users.request_sms(authy_id, {'force': True})

# if sms.ok():
#     print sms.content
# else:
#     print sms.errors()


# verification = authy_api.tokens.verify(authy_id, 2554557)

# print(verification.ok())
# print(verification.success)

tests/fakeauthy.py


import json

from threading import Thread

from BaseHTTPServer import BaseHTTPRequestHandler
from BaseHTTPServer import HTTPServer


class AuthyAPIHandler(BaseHTTPRequestHandler):

    requests = []
    responses = []

    default_response = ({}, 200, None)

    def do_GET(self):

        content_len = int(self.headers.getheader('content-length', 0))
        content = self.rfile.read(content_len)

        self.requests.append((self.path, content))

        try:
            response, status_code, headers = self.responses.pop(0)
        except IndexError:
            response, status_code, headers = self.default_response

        self.send_response(status_code)

        headers = headers or {'Content-Type': 'application/json'}
        for key, value in headers.items():
            self.send_header(key, value)
        self.end_headers()

        self.wfile.write(json.dumps(response))
        self.finish()

    do_POST = do_GET


class FakeAuthyAPIServer(object):

    def __init__(self, port, host='localhost'):

        self.host = host
        self.port = port
        self.server = None
        self.thread = None

    def set_response(self, response, code=200, headers=None):

        AuthyAPIHandler.responses.append((response, code, headers))

    def reset(self):

        AuthyAPIHandler.requests = []
        AuthyAPIHandler.responses = []

    def start(self):

        self.server = HTTPServer((self.host, self.port), AuthyAPIHandler)
        self.thread = Thread(target=self.server.serve_forever)
        self.thread.daemon = True
        self.thread.start()

    def stop(self):

        self.reset()
        self.server.shutdown()
        self.server.socket.close()
        self.thread = None


if __name__ == '__main__':

    import requests

    fake_authy_api = FakeAuthyAPIServer(port=12345, host='0.0.0.0')

    try:
        fake_authy_api.start()
        fake_authy_api.set_response({'message': 'Not Found'}, code=404)

        response = requests.get('http://localhost:12345/protected/json/phones/info',
                                headers={'Header': 'value'})
        print(response.status_code)
        print(response.text)

        resp = requests.get('http://localhost:12345/protected/json/phones/info',
                            headers={'Header': 'value'})
        print(response.status_code)
        print(response.text)
    finally:
        fake_authy_api.stop()


# EOF


tests/test_users.py
import six
import sys
import test_helper
import unittest

if six.PY3:
    from unittest.mock import MagicMock
else:
    from mock import MagicMock

from authy import AuthyException
from authy.api.resources import User, Sms
from authy.api.resources import Users
from authy.api import AuthyApiClient


class UsersTest(unittest.TestCase):

    def setUp(self):
        self.users = MagicMock(Users(test_helper.API_URL, test_helper.API_KEY))
        self.response = MagicMock()
        
        user = MagicMock(User(self.users, self.response))
        user.content = {"user": {"id": test_helper.API_USER_ID}, "success": True, "message": "something"}
        user.errors = MagicMock(return_value={})
        user.ok = MagicMock(return_value=True)
        user.id = MagicMock(return_value=test_helper.API_USER_ID)

        sms = MagicMock(Sms(self.users, self.response))
        sms.errors = MagicMock(return_value={})

        self.users.create = MagicMock(return_value = user)
        self.users.delete = MagicMock(return_value = user)
        self.users.status = MagicMock(return_value = user)
        self.users.request_sms = MagicMock(return_value = sms)

    def test_users(self):
        self.assertIsInstance(self.users, Users)

    def test_create_valid_user(self):
        user = self.users.create('test@example.com', '3457824988', 1)
        self.assertEqual(user.errors(), {})
        self.assertIsInstance(user, User)
        self.assertTrue(user.ok())
        self.assertTrue('user' in user.content)
        self.assertIsNotNone(user.id)

    def test_request_sms_token(self):
        user = self.users.create('test@example.com', '202-555-0158', 1)
        sms = self.users.request_sms(user.id)
        self.assertTrue(sms.ok())
        self.assertTrue(sms.content['success'])
        self.assertEqual(sms.errors(), {})
        self.assertEqual(user.errors(), {})

    def test_sms_ignored(self):
        user = self.users.create('test@example.com', '202-555-0197', 1)
        sms = self.users.request_sms(user.id)
        self.assertTrue(sms.ok())
        # fake 'ignored' field in JSON response
        sms.content['ignored'] = 'true'
        self.assertTrue(sms.ignored())

    def test_get_user_status(self):
        user = self.users.create('test@example.com', '3107810860', 1)
        status = self.users.status(user.id)
        self.assertTrue(status.ok(), msg="errors: {0}".format(status.errors()))
        self.assertTrue(status.content['success'])
        self.assertEqual(status.errors(), {})

    def test_delete_user(self):
        user = self.users.create('test@example.com', '3107810860', 1)
        user = self.users.delete(user.id)
        self.assertTrue(user.ok())
        self.assertTrue(user.content['success'])
        self.assertEqual(user.errors(), {})

if __name__ == "__main__":
	    unittest.main()


notes.sh
#!/bin/bash

REPOSITORY_URL="https://github.com/authy/authy-python"

PREVIOUS_TAG=$(git describe --abbrev=0 --tags)
NEW_VERSION=$(bumpversion --dry-run --list patch | grep new_version | sed s,"^.*=",,)
LATEST_TAG="v$NEW_VERSION"

# Get a log of commits that occured between two tags
# We only get the commit hash so we don't have to deal with a bunch of ugly parsing
# See Pretty format placeholders at https://git-scm.com/docs/pretty-formats
echo "Grabbing all commits since $PREVIOUS_TAG"
COMMITS=($(git log $PREVIOUS_TAG..HEAD --pretty=format:"%H"))

# Store our changelog in a variable to be saved to a file at the end
MARKDOWN="\nVersion $NEW_VERSION\n"
MARKDOWN+="--------------------\n"
MARKDOWN+="[Full Changelog]($REPOSITORY_URL/compare/$PREVIOUS_TAG...$LATEST_TAG)\n"

echo "Looking through ${#COMMITS[@]} commits for merged PRs"

# Loop over each commit and look for merged pull requests
for COMMIT in "${COMMITS[@]}"; do
	# Get the subject of the current commit
	SUBJECT=$(git log -1 ${COMMIT} --pretty=format:"%s")

	# If the subject contains "Merge pull request #xxxxx" then it is deemed a pull request
	PULL_REQUEST=$( grep -Eo "Merge pull request #[[:digit:]]+" <<< "$SUBJECT" )
  echo $PULL_REQUEST
	if [[ $PULL_REQUEST ]]; then
		# Perform a substring operation so we're left with just the digits of the pull request
		PULL_NUM=${PULL_REQUEST#"Merge pull request #"}

		# Get the body of the commit
		BODY=$(git log -1 ${COMMIT} --pretty=format:"%b")
		MARKDOWN+="\n - [#$PULL_NUM]($REPOSITORY_URL/pull/$PULL_NUM): $BODY"
  fi
done

echo "Updating CHANGES with merged PRs"

# Insert our new changes into the Changelog
awk -v n=5 -v s="$MARKDOWN" 'NR == n {print s} {print}' CHANGES > CHANGES.new

head CHANGES.new

mv CHANGES.new CHANGES

git commit -am "Log CHANGES for release $LATEST_TAG"


release.sh
#!/bin/bash

set -e

# make sure origin/master is what we expect
ORIGIN=$(git config --get remote.origin.url)
EXPECTED_ORIGIN="git@github.com:authy/authy-python.git"

if [ "$ORIGIN" != "$EXPECTED_ORIGIN" ]
then
  echo "Unexpected remote $ORIGIN. Please make sure your remote origin is $EXPECTED_ORIGIN"
  exit
else
  echo "Remote origin verified. Proceeding to get the latest from master."
fi

# make sure you have the bumpversion dependency
command -v bumpversion >/dev/null 2>&1 || { echo "Activate your virtualenv for 'bumpversion' dependency" >&2; exit 1; }

# this will fail if there are local changes. that's ok.
git checkout master
git fetch && git rebase

########################
# UPDATE RELEASE NOTES #
########################

echo "Updating release notes"

bash notes.sh

################
# BUMP VERSION #
################

# https://github.com/peritus/bumpversion
# This command does several things!
#    * updates patch versions in .bumpversion.cfg and authy/__init__.py
#    * Adds+commits to Git .bumpversion.cfg and authy/__init__.py files
#    * Adds Git tag with new version

echo 'Bumping patch version'

bumpversion patch

##################
# DEPLOY CHANGES #
##################

echo 'Commiting changes and pushing to master'

# commit changes to master
# this will kick off the Travis CI build which auto-deploys to PyPI
git push origin master --tags


authy/api/resources.py
import requests
import platform
import six
import hashlib
import hmac
import base64
import re

from authy import __version__, AuthyFormatException


try:
    from urllib import quote
except ImportError:
    from urllib.parse import quote

try:
    import json  # python 2.6+
except ImportError:
    import simplejson as json  # python 2.5

MIN_TOKEN_SIZE = 6
MAX_TOKEN_SIZE = 12
MAX_STRING_SIZE = 200

try:
    _PLAT_STR = platform.platform(True)
except IOError:
    _PLAT_STR = "unknown_platform"


class Resource(object):
    """
    Request CRUD defined in Resource.
    """

    def __init__(self, api_uri, api_key):
        """
        Class constructor
        :param string api_uri:
        :param string api_key:
        """
        self.api_uri = api_uri
        self.api_key = api_key
        self.def_headers = self.__default_headers()

    def post(self, path, data={}):
        """
        initiate create request to client server.
        :param string path:
        :param dict data Receives from child class:
        :return json response:
        """
        return self.request("POST", path, data, {'Content-Type': 'application/json'})

    def get(self, path, data={}):
        """
        initiate get request to client server.
        :param string path:
        :param dict data:
        :return Json response:
        """
        return self.request("GET", path, data)

    def put(self, path, data={}):
        """
        initiate update request.
        :param string path:
        :param dict data:
        :return:
        """
        return self.request("PUT", path, data, {'Content-Type': 'application/json'})

    def delete(self, path, data={}):
        """
        initiate recored delete request.
        :param string path:
        :param dict data:
        :return:
        """
        return self.request("DELETE", path, data)

    def request(self, method, path, data={}, headers={}):
        """

        :param callback method:
        :param string path:
        :param dict data:
        :param headers:
        :return:
        """
        url = self.api_uri + path
        params = {}
        temp = headers.copy()
        if six.PY3:
            temp = dict(
                self.def_headers.items() |
                headers.items()
            )
        else:
            temp = dict(
                self.def_headers.items() +
                headers.items()
            )
        headers = temp
        headers['X-Authy-API-Key'] = self.api_key
        if method == "GET":
            params.update(data)
            return requests.request(method, url, headers=headers, params=params)
        else:
            return requests.request(method, url, headers=headers, params=params, data=json.dumps(data))

    def __default_headers(self):
        return {
            'User-Agent': "AuthyPython/{0} ({1}; Python {2})".format(
                __version__,
                _PLAT_STR,
                platform.python_version()
            )}


class Instance(object):
    """
    Response parsing from resource request api calls
    """

    def __init__(self, resource, response):
        """
        Class constructor.
        :param class instance resource:
        :param class instance response:
                """
        self.resource = resource
        self.response = response

        try:
            self.content = self.response.json()
        except ValueError:
            self.content = self.response.text

    def ok(self):
        """

        :return Response True if success:
        """
        return self.response.status_code == 200

    def errors(self):
        """
        :return error dict if no success:
        """
        if self.ok():
            return {}

        errors = self.content

        if(not isinstance(errors, dict)):
            errors = {"error": errors}  # convert to dict for consistency
        elif('errors' in errors):
            errors = errors['errors']

        return errors

    def __getitem__(self, key):
        return self.content[key]


class Sms(Instance):
    """
    sms response handler
    """

    def ignored(self):
        try:
            self.content['ignored']
            return True
        except KeyError:
            return False


class Call(Instance):
    """
    call response handler
    """

    def ignored(self):
        try:
            self.content['ignored']
            return True
        except KeyError:
            return False


class User(Instance):
    """
    users class response handler.
    """

    def __init__(self, resource, response):
        """
        user constructor. assigns id if exist.
        :param resource class instance:
        :param response class instance:
        """
        super(User, self).__init__(resource, response)
        if(isinstance(self.content, dict) and 'user' in self.content):
            self.id = self.content['user']['id']
        else:
            self.id = None


class Users(Resource):
    """
    create, check status or delete user through this users datatype
    """

    def create(self, email, phone, country_code=1, send_install_link_via_sms=False):
        """
        sends request to create new user.
        :param string email:
        :param string phone:
        :param string country_code:
        :param bool send_install_link_via_sms:
        :return:
        """
        data = {
            "user": {
                "email": email,
                "cellphone": phone,
                "country_code": country_code
            },

            'send_install_link_via_sms': send_install_link_via_sms
        }

        resp = self.post("/protected/json/users/new", data)

        return User(self, resp)

    def request_sms(self, user_id, options={}):
        resp = self.get("/protected/json/sms/" + quote(str(user_id)), options)

        return Sms(self, resp)

    def request_call(self, user_id, options={}):
        resp = self.get("/protected/json/call/" + quote(str(user_id)), options)

        return Call(self, resp)

    def status(self, user_id):
        resp = self.get("/protected/json/users/{0}/status".format(user_id))

        return User(self, resp)

    def delete(self, user_id):
        resp = self.post("/protected/json/users/{0}/delete".format(user_id))

        return User(self, resp)
    
    def generate_qr(self, user_id, size=None, label=None):
        data = {}

        if size is not None:
            data['qr_size'] = size
        if label is not None:
            data['label'] = label

        resp = self.post("/protected/json/users/{}/secret".format(user_id), data)

        return User(self, resp)

    def registration_status(self, custom_user_id):
        resp = self.get("/protected/json/registrations/status?custom_user_id={0}".format(custom_user_id))

        return User(self, resp)


class Token(Instance):

    def ok(self):
        if super(Token, self).ok():
            return '"token":"is valid"' in str(self.response.content)
        return False


class Tokens(Resource):

    def verify(self, device_id, token, options={}):
        self.__validate(token, device_id)
        if 'force' not in options:
            options['force'] = "true"
        resp = self.get(
            "/protected/json/verify/{0}/{1}".format(quote(str(token)), quote(str(device_id))), options)
        return Token(self, resp)

    def __validate(self, token, device_id):
        self.__validate_digit(token, "Invalid Token. Only digits accepted.")
        self.__validate_digit(
            device_id, "Invalid Authy id. Only digits accepted.")
        length = len(str(token))
        if length < MIN_TOKEN_SIZE or length > MAX_TOKEN_SIZE:
            raise AuthyFormatException("Invalid Token. Unexpected length.")

    def __validate_digit(self, var, message):
        if six.PY3:
            if not isinstance(var, (int)) and not var.isdigit():
                raise AuthyFormatException(message)
        else:
            if not isinstance(var, (int, long)) and not var.isdigit():
                raise AuthyFormatException(message)


class App(Instance):
    pass


class Apps(Resource):

    def fetch(self):
        resp = self.get("/protected/json/app/details")
        return App(self, resp)


class Stats(Instance):
    pass


class StatsResource(Resource):

    def fetch(self):
        resp = self.get("/protected/json/app/stats")
        return Stats(self, resp)


class Phone(Instance):
    pass


class Phones(Resource):

    def __validate_channel(self, via):
        if via != 'sms' and via != 'call':
            raise AuthyFormatException("Invalid Via. Expected 'sms' or 'call'.")
    
    def __validate_code_length(self, code_length):
        try:
            cl = int(code_length)
            if cl < 4 or cl > 10:
                raise ValueError
            return cl
        except ValueError:
            raise AuthyFormatException(
                "Invalid code_length. Expected numeric value from 4-10.")

    def verification_start(self, phone_number, country_code, via='sms',
                           locale=None, code_length=4):
        """
        :param string phone_number: stored in your databse or you provided while creating new user.
        :param string country_code: stored in your databse or you provided while creating new user.
        :param string via: verification method either sms or call
        :param string locale: optional default none
        :param number code_length: optional default 4
        :return:
        """

        self.__validate_channel(via)

        options = {
            'phone_number': phone_number,
            'country_code': country_code,
            'via': via
        }

        cl = self.__validate_code_length(code_length)
        options['code_length'] = cl

        if locale:
            options['locale'] = locale

        resp = self.post("/protected/json/phones/verification/start", options)
        return Phone(self, resp)

    def verification_check(self, phone_number, country_code, verification_code):
        """
        :param phone_number:
        :param country_code:
        :param verification_code:
        :return:
        """
        options = {
            'phone_number': phone_number,
            'country_code': country_code,
            'verification_code': verification_code
        }
        resp = self.get("/protected/json/phones/verification/check", options)
        return Phone(self, resp)

    def info(self, phone_number, country_code):
        options = {
            'phone_number': phone_number,
            'country_code': country_code
        }
        resp = self.get("/protected/json/phones/info", options)
        return Phone(self, resp)


class OneTouchResponse(Instance):
    """
    OneTouch response handler.
    """

    def __init__(self, resource, response):
        """
        OneTouchResponse constructor. receives two prams
        :param resource instance:
        :param response of OneTouch datatype:
        """
        self.uuid = None
        super(OneTouchResponse, self).__init__(resource, response)
        if (isinstance(self.content, dict) and 'approval_request' in self.content):
            self.uuid = self.content['approval_request']['uuid']

    def get_uuid(self):
        if not self.uuid:
            return False
        return self.uuid

    def status(self):
        success = False

        if ('success' in self.content):
            success = self.content['success']

        return success


class OneTouch(Resource):

    def _validate_request(self, user_id, message):
        if not user_id or not isinstance(user_id, int):
            raise AuthyFormatException(
                'Invalid authy id, user id is requred and must be an integer value.')

        if not message:
            raise AuthyFormatException(
                'Invalid message - should not be empty. It is required')

    def send_request(self, user_id, message, seconds_to_expire=None, details={}, hidden_details={}, logos=[]):
        """
        OneTouch verification request. Sends a request for Auth App. For more info https://www.twilio.com/docs/api/authy/authy-onetouch-api
        :param string user_id: user_id User's authy id stored in your database
        :param string message: Required, the message shown to the user when the approval request arrives.
        :param number seconds_to_expire: Optional, defaults to 120 (two minutes).
        :param dict details:  For example details['Requested by'] = 'MacBook Pro, Chrome'; it will be displayed on Authy app
        :param dict hidden_details: Same usage as detail except this detail is not shown in Authy app
        :param list logos: Contains the logos that will be shown to user. The logos parameter is expected to be an array of objects, each object with two fields: res (values are default,low,med,high) and url
        :return OneTouchResponse: the server response Json Object
        """

        self._validate_request(user_id, message)

        data = {
            "message": message[:MAX_STRING_SIZE],
            "seconds_to_expire": seconds_to_expire,
            "details": self.__clean_inputs(details),
            'hidden_details': self.__clean_inputs(hidden_details),
            'logos': self.clean_logos(logos)
        }

        request_url = "/onetouch/json/users/{0}/approval_requests".format(
            user_id)
        response = self.post(request_url, data)
        return OneTouchResponse(self, response)

    def clean_logos(self, logos):
        """
        Validate logos input.
        :param list logos:
        :return list logos:
        """
        if not len(logos):
            return logos  # Allow nil hash
        if not isinstance(logos, list):
            raise AuthyFormatException(
                'Invalid logos list. Only res and url required')

        temp_array = {}
        clean_logos = []

        for logo in logos:

            if not isinstance(logo, dict):
                raise AuthyFormatException('Invalid logo type')

            for l in logo:
                # We ignore any additional parameter on the logos, and truncate
                # string size to the maximum allowed.
                if l == 'res':
                    temp_array['res'] = logo[l][:MAX_STRING_SIZE]
                elif l == 'url':
                    temp_array['url'] = logo[l][:MAX_STRING_SIZE]
                else:
                    raise AuthyFormatException(
                        'Invalid logos list. Only res and url required')

            clean_logos.append(temp_array)
            temp_array = {}

        return clean_logos

    def get_approval_status(self, uuid):
        """
        OneTouch verification request. Sends a request for Auth App. For more info https://www.twilio.com/docs/api/authy/authy-onetouch-api
        :param string uuid Required. The approval request ID. (Obtained from the response to an ApprovalRequest):
        :return OneTouchResponse the server response Json Object:
        """
        request_url = "/onetouch/json/approval_requests/{0}".format(uuid)
        response = self.get(request_url)
        return OneTouchResponse(self, response)

    def validate_one_touch_signature(self, signature, nonce, method, url, params):
        """
        Function to validate signature in X-Authy-Signature key of headers.

        :param string signature: X-Authy-Signature key of headers.
        :param string nonce: X-Authy-Signature-Nonce key of headers.
        :param string method: GET or POST - configured in app settings for OneTouch.
        :param string url: base callback url.
        :param dict params: params sent by Authy.
        :return bool: True if calculated signature and X-Authy-Signature are identical else False.
        """
        if not signature or not isinstance(signature, str):
            raise AuthyFormatException(
                "Invalid signature - should not be empty. It is required")

        if not nonce:
            raise AuthyFormatException(
                "Invalid nonce - should not be empty. It is required")

        if not method or not ('get' == method.lower() or 'post' == method.lower()):
            raise AuthyFormatException(
                "Invalid method - should not be empty. It is required")

        if not params or not isinstance(params, dict):
            raise AuthyFormatException(
                "Invalid params - should not be empty. It is required")

        query_params = self.__make_http_query(params)
        # Sort and replace encoded  params in case-sensitive order
        sorted_params = '&'.join(sorted(query_params.replace(
            '/', '%2F').replace('%20', '+').split('&')))
        sorted_params = re.sub("\\%5B([0-9])*\\%5D", "%5B%5D", sorted_params)
        sorted_params = re.sub("\\=None", "=", sorted_params)
        data = nonce + "|" + method + "|" + url + "|" + sorted_params
        try:
            calculated_signature = base64.b64encode(
                hmac.new(self.api_key.encode(), data.encode(), hashlib.sha256).digest())
            return calculated_signature.decode() == signature
        except:
            calculated_signature = base64.b64encode(
                hmac.new(self.api_key, data, hashlib.sha256).digest())
            return calculated_signature == signature

    def __make_http_query(self, params, topkey=''):
        """
        Function to covert params into url encoded query string
        :param dict params: Json string sent  by Authy.
        :param string topkey: params key
        :return string: url encoded Query.
        """
        if len(params) == 0:
            return ""
        result = ""
        # is a dictionary?
        if type(params) is dict:
            for key in params.keys():
                newkey = quote(key)
                if topkey != '':
                    newkey = topkey + quote('[' + key + ']')
                if type(params[key]) is dict:
                    result += self.__make_http_query(params[key], newkey)
                elif type(params[key]) is list:
                    i = 0
                    for val in params[key]:
                        if type(val) is dict:
                            result += self.__make_http_query(
                                val, newkey + quote('['+str(i)+']'))
                        else:
                            result += newkey + \
                                quote('['+str(i)+']') + "=" + \
                                quote(str(val)) + "&"
                        i = i + 1
                # boolean should have special treatment as well
                elif type(params[key]) is bool:
                    result += newkey + "=" + \
                        quote(str(params[key]).lower()) + "&"
                # assume string (integers and floats work well)
                else:
                    result += newkey + "=" + quote(str(params[key])) + "&"
        # remove the last '&'
        if (result) and (topkey == '') and (result[-1] == '&'):
            result = result[:-1]
        return result

    def __clean_inputs(self, params):

        if not isinstance(params, dict):
            return params[:MAX_STRING_SIZE]

        temp_hash = {}

        for k in params:
            if not isinstance(params[k], dict):
                temp_hash[k[:MAX_STRING_SIZE]] = params[k][:MAX_STRING_SIZE]
            else:
                temp_hash[k[:MAX_STRING_SIZE]] = self.__clean_inputs(params[k])

        return temp_hash
